### PR TITLE
Upgrade tests

### PR DESCRIPTION
**Pull request by ianmacclancy**
# Description
Adding upgrade tests that will run on each PR
Issue: https://github.com/solo-io/solo-projects/issues/4189
Stability Milestone: https://github.com/solo-io/solo-projects/milestone/18 

# Context
Tests will be 
- Last minor latests patch release -> PR code 
- Current Minor last patch release -> PR code

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -739,13 +739,12 @@ docker-push-local-arm: docker docker-push
...
 # This prevents executing the dependent targets if CREATE_ASSETS is not true, while still enabling `make docker`
 # to be used for local testing.
-# docker-push-non-arm is intended to be run on CI only, where as docker-push-local is inteneded for local builds. Primarily used for arm support.
+# docker-push-non-arm is intended to be run on CI only, where as docker-push-local is intended for local builds. Primarily used for arm support.
 .PHONY: docker-push
 docker-push: docker-push-local docker-push-non-arm
 
 .PHONY: docker-push-local
 docker-push-local: $(DOCKER_IMAGES)
-ifeq ($(CREATE_ASSETS), "true")
```
Comment (jenshu) [inline Makefile:748]: why was this removed?

File: .github/workflows/regression-tests.yaml [hunk 2]
```diff
@@ -139,6 +139,7 @@ jobs:
       if: needs.prepare_env.outputs.should-pass-regression-tests != 'true'
       env:
         KUBE2E_TESTS: ${{ matrix.kube-e2e-test-type }}
+        GITHUB_TOKEN: ${{ github.token }}
```
Comment (jenshu) [inline .github/workflows/regression-tests.yaml:142]: why did we need to add this?

File: test/kube2e/upgrade/upgrade_test.go [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+		Context("When upgrading from LastPatchMostRecentMinorVersion to PR version of gloo", func() {
+			BeforeEach(func() {
+				installGloo(testHelper, LastPatchMostRecentMinorVersion.String(), strictValidation)
+			})
+			AfterEach(func() {
+				uninstallGloo(testHelper, ctx, cancel)
+			})
+			//It("Used for local testing to check base case upgrades", func() {
+			//	baseUpgradeTest(ctx, crdDir, LastPatchMostRecentMinorVersion.String(), testHelper, chartUri, strictValidation)
+			//})
```
Comment (jenshu) [inline test/kube2e/upgrade/upgrade_test.go:97]: should this be removed or uncommented?

File: test/kube2e/upgrade/upgrade_test.go [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+			})
+		})
+	})
+})
+
+// ===================================
+// Repeated Test Code
+// ===================================
+// Based case test for local runs to help narrow down failures
+func baseUpgradeTest(ctx context.Context, crdDir string, startingVersion string, testHelper *helper.SoloTestHelper, chartUri string, strictValidation bool) {
```
Comment (jenshu) [inline test/kube2e/upgrade/upgrade_test.go:164]: much of this code is duplicated from the helm tests, can we refactor them out into a shared pkg?

File: test/kube2e/upgrade/upgrade_test.go [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+)
+
+const namespace = defaults.GlooSystem
+
+var _ = Describe("Kube2e: Upgrade Tests", func() {
```
Comment (jenshu) [inline test/kube2e/upgrade/upgrade_test.go:44]: does this suite share any tests in common with the helm suite? should we remove any redundant tests (from either here or the helm suite)?

File: test/kube2e/upgrade/upgrade_utils.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+	var version1 = *a[i]
+	var version2 = *a[j]
+	return version2.MustIsGreaterThan(version1)
+}
+func (a ByVersion) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+
+func main() {
+	ctx := context.Background()
+	GetUpgradeVersions(ctx)
+}
```
Comment (jenshu) [inline test/kube2e/upgrade/upgrade_utils.go:37]: is this main function needed?

File: test/kube2e/upgrade/upgrade_utils.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
+package upgrade
```
Comment (jenshu) [inline test/kube2e/upgrade/upgrade_utils.go:1]:
```suggestion
package upgrade_test
```

File: test/kube2e/upgrade/upgrade_utils.go [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+	return version2.MustIsGreaterThan(version1)
+}
+func (a ByVersion) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+
+func main() {
+	ctx := context.Background()
+	GetUpgradeVersions(ctx)
+}
+
+func GetUpgradeVersions(ctx context.Context) (lastMinorLatestPatchVersion *versionutils.Version, currentMinorLatestPatchVersion *versionutils.Version, err error) {
```
Comment (jenshu) [inline test/kube2e/upgrade/upgrade_utils.go:39]: these functions can all be lowercased as they're only used within this pkg or file

File: Makefile (hunk 1 above)
Comment (ianmacclancy) [reply Makefile:748]: This check was preventing docker images pushing for arm builds. It is not used when we run ci and will enable people with m1s

File: test/kube2e/upgrade/upgrade_test.go (hunk 5 above)
Comment (ianmacclancy) [reply test/kube2e/upgrade/upgrade_test.go:44]: They share similar tests but not the same. In a future pr I am going to add the 1.12 specific upgrade tests from the helm test suite and refactor the code base to clean up shared code where I can. For now I want to keep this pr simpler and do the refactoring of helm in a separate pr

File: test/kube2e/upgrade/upgrade_test.go (hunk 3 above)
Comment (ianmacclancy) [reply test/kube2e/upgrade/upgrade_test.go:97]: Nope! I added the description to help clarify why this commented code is left in. Its useful for debugging to have an upgrade test that only upgrades and does nothing else. This can help debug upgrade issues. It doe not need to run for ci as its covered by the other test cases

File: test/kube2e/upgrade/upgrade_test.go (hunk 4 above)
Comment (ianmacclancy) [reply test/kube2e/upgrade/upgrade_test.go:164]: See comment above

File: test/kube2e/upgrade/upgrade_utils.go (hunk 8 above)
Comment (ianmacclancy) [reply test/kube2e/upgrade/upgrade_utils.go:39]: They will be used in solo-projects

File: test/kube2e/upgrade/upgrade_utils.go (hunk 6 above)
Comment (ianmacclancy) [reply test/kube2e/upgrade/upgrade_utils.go:37]: Nope, testing artifact. I'll remove it!

File: Makefile (hunk 1 above)
Comment (nfuden) [reply Makefile:748]: Since we've been chatting about separation of concerns do you think we might be able to smash this up as a separate first pr?

Comment (ianmacclancy) [reply Makefile:748]: Sure!

File: test/kube2e/upgrade/upgrade_utils.go (hunk 7 above)
Comment (ianmacclancy) [reply test/kube2e/upgrade/upgrade_utils.go:1]: The file ending is not _test so it cannot be a test package file

File: test/kube2e/upgrade/upgrade_utils.go [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+	}
+	lastMinorLatestPatchVersion, lastMinorErr := GetLatestReleasedVersion(ctx, currentMinorLatestPatchVersion.Major, currentMinorLatestPatchVersion.Minor-1)
+	if lastMinorErr != nil {
+		return nil, nil, lastMinorErr
+	}
+	return lastMinorLatestPatchVersion, currentMinorLatestPatchVersion, curMinorErr
+}
+
+func GetLastReleaseOfCurrentMinor() (*versionutils.Version, error) {
+	repo_name := "gloo"                    // pull out to const
```
Comment (jenshu) [inline test/kube2e/upgrade/upgrade_utils.go:49]: if this will be used by solo-projects should the repoName be an arg to this function?

Comment (ianmacclancy) [reply test/kube2e/upgrade/upgrade_utils.go:49]: This will come out of the refactor. It makes more sense to leave them exported and then continue the work in the next PR thats simplifying all code for re-used and pulling out the shared code. I expect there to be flux in a bunch of the upgrade_util methods.

Comment (elcasteel) [reply test/kube2e/upgrade/upgrade_utils.go:49]: In the interest of the principle of "avoid breaking changes," I think we shouldn't export these methods until they have a signature that we want to export. Which is to say either figuring out how to get the repo name now or lowercasing the names until we're ready.

**Review by elcasteel — APPROVED** [review]

File: test/kube2e/upgrade/upgrade_utils.go (hunk 9 above)
Comment (ianmacclancy) [reply test/kube2e/upgrade/upgrade_utils.go]: Updated

**Review by jenshu — APPROVED** [review]
